MON-4133: add unknown content type as a reason for TargetDown and make markdownlint happy (#232)
openshift/runbooks: alerts/cluster-monitoring-operator/TargetDown.md

workflow.

### Potential Issues and Resolutions

#### Unknown Content Type

Beginning with OCP `4.19`, Prometheus is more strict concerning the `Content-Type`
header received when scraping, if the target being scraped does not specify
a `Content-Type` header or if the header is unparsable or unrecognised and
no [`fallback_scrape_protocol`] is set for the target, the scrape is considered
as having failed, which would lead to the alert being fired.

For more context, error logs similar to the one below should be present:

   ```shell
   $ NAMESPACE='<value of namespace label from alert>'

   $ oc -n $NAMESPACE logs -l 'app.kubernetes.io/name=prometheus' --tail=-1 | \
   grep 'Failed to determine correct type of scrape target.*' \
   | sort | uniq -c | sort -n
   msg="Failed to determine correct type of scrape target." \
   component="scrape manager" \
   scrape_pool=<monitor>/<namespace>/<service-monitor-name>/<endpoint-id> \
   target=concerned-target content_type="concerned-content_type" \
   fallback_media_type="fallback_media_type-if-any" err="explicit_error"
   ```

> In such a case, please open a support case.

Author: maybe exclude `foo-bar` namespaces with unjustified `openshift.io/cluster-monitoring=true`?
Or maybe that could be done by support?

workflow.

### Potential Issues and Resolutions

#### Unknown Content Type

Beginning with OCP `4.19`, Prometheus is more strict concerning the `Content-Type`
header received when scraping, if the target being scraped does not specify
a `Content-Type` header or if the header is unparsable or unrecognised and
no [`fallback_scrape_protocol`] is set for the target, the scrape is considered

> as having failed, which would lead to the alert being fired.

Reviewer: `The scrape is considered as having failed`.. I wanted to make sure here: does this mean that Prometheus just perceives it  as failed, or does it really fail?

Author: Prometheus will not scrape the target, metrics are not retrieved from that target.

> Beginning with OCP `4.19`, Prometheus is more strict concerning the `Content-Type`
> header received when scraping, if the target being scraped does not specify
> a `Content-Type` header or if the header is unparsable or unrecognised and
> no [`fallback_scrape_protocol`] is set for the target, the scrape is considered
> as having failed, which would lead to the alert being fired.

Reviewer: ```suggestion
Beginning with OpenShift Container Platform version 4.19, Prometheus enforces stricter rules on the `Content-Type` header when scraping.
If the scraped target does not have a `Content-Type` header or if the header is unparsable or unrecognised and no [`fallback_scrape_protocol`] is set for the target, the scrape fails, and the `TargetDown` alert is fired.
```

Reviewer: here I assumed that the scrape fails (maybe even if it is just considered as failed, it would still be okay to just say that it failed (I mean, from the user perspective, it might very well be the same thing :D )

Author: You're right, clearer, I'll change.

> For more context, error logs similar to the one below should be present:

Reviewer: ```suggestion
For more context, error logs similar to the following one should be present:
```

   ```shell

>    $ NAMESPACE='<value of namespace label from alert>'

Reviewer: I believe this should be a comment

>    $ NAMESPACE='<value of namespace label from alert>'

Author: why not

   $ oc -n $NAMESPACE logs -l 'app.kubernetes.io/name=prometheus' --tail=-1 | \
   grep 'Failed to determine correct type of scrape target.*' \
   | sort | uniq -c | sort -n
   msg="Failed to determine correct type of scrape target." \
   component="scrape manager" \
   scrape_pool=<monitor>/<namespace>/<service-monitor-name>/<endpoint-id> \
   target=concerned-target content_type="concerned-content_type" \
   fallback_media_type="fallback_media_type-if-any" err="explicit_error"
   ```

> In such a case, please open a support case, attaching the artifacts gathered
> during the diagnosis procedure.

Reviewer: ```suggestion
In such a case, open a support case and attach the artifacts gathered
during the diagnosis procedure.
```

workflow.

### Potential Issues and Resolutions

#### Unknown Content Type

Beginning with OpenShift Container Platform version 4.19, Prometheus enforces stricter
rules on the `Content-Type` header when scraping.

If the scraped target does not have a `Content-Type` header or if the header is
unparsable or unrecognised and no [`fallback_scrape_protocol`] is set for the

> target, the scrape fails, and the `TargetDown` alert is fired.

Reviewer 2: ```suggestion
target, the scrape fails and the `TargetDown` alert is fired.
```

Author: That was suggested by https://github.com/openshift/runbooks/pull/232#discussion_r1954421745

For more context, error logs similar to the following one should be present:

   ```shell
   # NAMESPACE='<value of namespace label from alert>'

   $ oc -n $NAMESPACE logs -l 'app.kubernetes.io/name=prometheus' --tail=-1 | \
   grep 'Failed to determine correct type of scrape target.*' \
   | sort | uniq -c | sort -n
   msg="Failed to determine correct type of scrape target." \
   component="scrape manager" \
   scrape_pool=<monitor>/<namespace>/<service-monitor-name>/<endpoint-id> \
   target=concerned-target content_type="concerned-content_type" \
   fallback_media_type="fallback_media_type-if-any" err="explicit_error"
   ```

> In such a case, open a support case and attach the artifacts gathered

Reviewer 2: I think we should make it clear what it is meant by artifacts or add link to diagnosis steps from openshift docs

Author: We (still) use that wording in many other places: https://github.com/search?q=repo%3Aopenshift%2Frunbooks%20artifacts&type=code
We can always make clearer in the future if needed, for now let's stay consistent.

      app.kubernetes.io/name: node-exporter
      app.kubernetes.io/part-of: openshift-monitoring
```

In the `Prometheus` that scrapes this target, the volume mounted at `/etc/prometheus/secrets/*`
are the secrets having the same name as the subdirectory. In this example, the secret
is `metrics-client-certs` in the namespace `openshift-monitoring`. Now we extract
the certificate from the secret using this command:

```shell
oc extract secret/$SECRET_NAME -n $NAMESPACE --keys=tls.crt --to=- > certificate.crt
```

> Then we inspect is expiration date.

Reviewer 3: ```suggestion
Then we inspect its expiration date.
```

Author: out of scope of the PR (was there before the change), but done.
